executor/write.go

```diff
@@ -69,28 +71,30 @@ func updateRecord(ctx sessionctx.Context, h int64, oldData, newData []types.Datu
 			// Cast changed fields with respective columns.
 			v, err := table.CastValue(ctx, newData[i], col.ToInfo())
 			if err != nil {
-				return false, handleChanged, newHandle, errors.Trace(err)
+				return false, handleChanged, newHandle, 0, errors.Trace(err)
 			}
 			newData[i] = v
 		}
 
 		// Rebase auto increment id if the field is changed.
 		if mysql.HasAutoIncrementFlag(col.Flag) {
 			if newData[i].IsNull() {
-				return false, handleChanged, newHandle, errors.Errorf("Column '%v' cannot be null", col.Name.O)
+				return false, handleChanged, newHandle, 0,
+					errors.Errorf("Column '%v' cannot be null", col.Name.O)
 			}
 			val, errTI := newData[i].ToInt64(sc)
 			if errTI != nil {
-				return false, handleChanged, newHandle, errors.Trace(errTI)
+				return false, handleChanged, newHandle, 0, errors.Trace(errTI)
 			}
+			lastInsertID = uint64(val)
```

Comment:
What if the col is not primary key?

Reply:
I think there is no difference between them.

Reply:
Could it possible that the column is auto_increament but with unsign flag?

Reply:
Yes, it could be. The behavior of MySQL is strange. When doing duplicate update in insert statement, 

- if the assignment is `on duplicate update i = 10`, then `select last_insert_id()` returns 0;
- if the assignment is `on duplicate update i = last_insert_id(i)`, then `select last_insert_id()` returns the really last_insert_id, but it depends on the value of the last_insert_id.

For the second case, please try the following sqls on MySQL:
```
create table t (i int not null auto_increment primary key, j int unique key);
insert into t values (-10, 1);
select last_insert_id();          -- will return 0
insert into t (j) values (1) on duplicate key update i = last_insert_id(i);  -- will raise an an error
```

It seems MySQL does return 0 but not -10 when `select last_insert_id()`, but it still set -10 is the real last insert id in the next insert statement.